File: java/course_schedule.java
public class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        if (numCourses==0) return true;
        HashSet<Integer>[] graph=new HashSet[numCourses];
        for (int i=0; i<numCourses; i++) {
            graph[i]=new HashSet<Integer>();
        }
        int[] incomingCnt=new int[numCourses];
        
        for (int i=0; i<prerequisites.length; i++) {
            if (!graph[prerequisites[i][1]].contains(prerequisites[i][0])) {
                graph[prerequisites[i][1]].add(prerequisites[i][0]);
                incomingCnt[prerequisites[i][0]]++;
            }
        }
        
        TreeSet<Integer> set=new TreeSet<Integer>();
        for (int i=0; i<incomingCnt.length; i++) {
            if (incomingCnt[i]==0) set.add(i);
        }
        
        while (!set.isEmpty()) {
            int n=set.pollFirst();
            
            for (int neighbor:graph[n]) {
                incomingCnt[neighbor]--;
                if (incomingCnt[neighbor]==0) {
                    set.add(neighbor);
                }
            }
            graph[n].clear();
        }
        
        for (int i=0; i<numCourses; i++) {
            if (!graph[i].isEmpty()) return false;
        }
        return true;
    }
}

File: python/course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph=[[] for i in range(numCourses)]
        incomingCnt=[0]*numCourses
        for i, j in prerequisites:
            if i not in graph[j]:
                graph[j].append(i)
                incomingCnt[i]+=1
                
        st=[i for i in range(numCourses) if incomingCnt[i]==0]
        while st:
            f=st.pop(0)
            for neighbor in graph[f]:
                incomingCnt[neighbor]-=1
                if incomingCnt[neighbor]==0:
                    st.append(neighbor)
                    
        for i in range(numCourses):
            if incomingCnt[i]>0: return False
        return True

